airbyte-integrations/bigquery-destination/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java

```diff
@@ -0,0 +1,329 @@
+
+      return BigQueryOptions.newBuilder()
+          .setProjectId(projectId)
+          .setCredentials(credentials)
+          .build()
+          .getService();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  static ImmutablePair<Job, String> executeQuery(BigQuery bigquery, QueryJobConfiguration queryConfig) {
+    JobId jobId = JobId.of(UUID.randomUUID().toString());
+    Job queryJob = bigquery.create(JobInfo.newBuilder(queryConfig).setJobId(jobId).build());
+    return executeQuery(queryJob);
+  }
+
+  private static ImmutablePair<Job, String> executeQuery(Job queryJob) {
+    final Job completedJob = waitForQuery(queryJob);
+    if (completedJob == null) {
+      throw new RuntimeException("Job no longer exists");
+    } else if (completedJob.getStatus().getError() != null) {
+      // You can also look at queryJob.getStatus().getExecutionErrors() for all
+      // errors, not just the latest one.
+      return ImmutablePair.of(null, (completedJob.getStatus().getError().toString()));
+    }
+
+    return ImmutablePair.of(completedJob, null);
+  }
+
+  private static Job waitForQuery(Job queryJob) {
+    try {
+      return queryJob.waitFor();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public StandardDiscoverSchemaOutput discover(JsonNode config) {
+    throw new RuntimeException("Not Implemented");
+  }
+
+  /**
+   * Strategy:
+   * <p>
+   * 1. Create a temporary table for each stream
+   * </p>
+   * <p>
+   * 2. Write records to each stream directly (the bigquery client handles managing when to push the
+   * records over the network)
+   * </p>
+   * <p>
+   * 4. Once all records have been written close the writers, so that any remaining records are
+   * flushed.
+   * </p>
+   * <p>
+   * 5. Copy the temp tables to the final table name (overwriting if necessary).
+   * </p>
+   *
+   * @param config - integration-specific configuration object as json. e.g. { "username": "airbyte",
+   *        "password": "super secure" }
+   * @param schema - schema of the incoming messages.
+   * @return consumer that writes singer messages to the database.
+   */
+  @Override
+  public DestinationConsumer<SingerMessage> write(JsonNode config, Schema schema) {
+
+    final BigQuery bigquery = getBigQuery(config);
+    Map<String, WriteConfig> writeConfigs = new HashMap<>();
+    final String datasetId = config.get("dataset_id").asText();
+
+    // create tmp tables if not exist
+    for (final Stream stream : schema.getStreams()) {
+      final String tableName = stream.getName();
+      final String tmpTableName = stream.getName() + "_" + Instant.now().toEpochMilli();
+
+      createTable(bigquery, datasetId, tmpTableName);
+      // https://cloud.google.com/bigquery/docs/loading-data-local#loading_data_from_a_local_data_source
+      final WriteChannelConfiguration writeChannelConfiguration = WriteChannelConfiguration
```

Comment:
note: worth explaining this non obvious choice here. I'm using this api which is designed for file uploads (not streaming uploads) that is a little harder to use, because inserts are free as opposed to the bq streaming API (which costs `$0.010 per 200 MB)` (that singer uses). 

> You are not charged for loading data from Cloud Storage or from local files into BigQuery. 

will add a comment that explains all this.

Reply:
this means that use of the connect for the user should be totally free. copy also is. free operation. the only charges are due to storage on disk.